homeassistant/components/androidtv/__init__.py

```diff
@@ -1 +1,195 @@
 """Support for functionality to interact with Android TV/Fire TV devices."""
+import json
+import logging
+import os
+
+from adb_shell.auth.keygen import keygen
+from androidtv import state_detection_rules_validator
+from androidtv.adb_manager.adb_manager_sync import ADBPythonSync
+from androidtv.setup_async import setup
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    CONF_DEVICE_CLASS,
+    CONF_HOST,
+    CONF_PORT,
+    EVENT_HOMEASSISTANT_STOP,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+from homeassistant.helpers.storage import STORAGE_DIR
+
+from .const import (
+    ANDROID_DEV,
+    ANDROID_DEV_OPT,
+    CONF_ADB_SERVER_IP,
+    CONF_ADB_SERVER_PORT,
+    CONF_ADBKEY,
+    CONF_STATE_DETECTION_RULES,
+    DEVICE_ANDROIDTV,
+    DEVICE_FIRETV,
+    DOMAIN,
+    MIGRATION_DATA,
+    SIGNAL_CONFIG_ENTITY,
+)
+
+PLATFORMS = ["media_player"]
+RELOAD_OPTIONS = [CONF_STATE_DETECTION_RULES]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def _setup_androidtv(hass, config):
+    """Generate an ADB key (if needed) and load it."""
+    adbkey = config.get(CONF_ADBKEY, hass.config.path(STORAGE_DIR, "androidtv_adbkey"))
+    if CONF_ADB_SERVER_IP not in config:
+        # Use "adb_shell" (Python ADB implementation)
+        if not os.path.isfile(adbkey):
+            # Generate ADB key files
+            keygen(adbkey)
+
+        # Load the ADB key
+        signer = ADBPythonSync.load_adbkey(adbkey)
+        adb_log = f"using Python ADB implementation with adbkey='{adbkey}'"
+
+    else:
+        # Use "pure-python-adb" (communicate with ADB server)
+        signer = None
+        adb_log = f"using ADB server at {config[CONF_ADB_SERVER_IP]}:{config[CONF_ADB_SERVER_PORT]}"
+
+    return adbkey, signer, adb_log
+
+
+def validate_state_det_rules(state_det_rules):
+    """Validate a string that contain state detection rules and return a dict."""
+    if not state_det_rules:
+        return None
+
+    try:
+        json_rules = json.loads(state_det_rules)
+    except ValueError:
+        _LOGGER.warning("Error loading state detection rules")
+        return None
+    try:
+        state_detection_rules_validator(json_rules, ValueError)
+    except ValueError as exc:
+        _LOGGER.warning("Invalid state detection rules: %s", exc)
+        return None
+    return json_rules
+
+
+async def async_connect_androidtv(
+    hass, config, *, state_detection_rules=None, timeout=10.0
```

Comment:
10.0 means that the user has 10 seconds to accept the connection on their Android device. It's fine if they don't accept it within that window, the next time it tries to connect the key will be pre-approved. But I think the config flow would fail. Should a larger timeout be used?

Reply:
This is already implemented. In the config flow the call to method `async_connect_androidtv` is done using `timeout=30.0`

Reply:
In that case, maybe the default value should be 30 instead of 10? When the key pair has already been approved, that parameter isn't used.

Reply:
In my mind it is better to set a lower default timeout and increase it when required. By example, if the `adbkey` is cancelled, we don't want to wait 30 sec if we are not in config flow. What do you think?

Reply:
I'm not sure what you mean by "if the `adbkey` is cancelled." Do you mean that the user clears the trusted keys on their Android device? Or if the user rejects the connection when the pop-up appears on their TV? The latter would be more likely to happen during config flow, but could happen on any reconnect attempt if the user didn't select "always trust connections from this computer."

Either way, I think it's an unlikely scenario. And if the user does something like that, I think it falls outside of the intended usage and having to wait 30 seconds instead of 10 is not unreasonable. 

I would view that timeout as

* If you're connecting for the first time (in config flow), you have 30 seconds to approve the connection or the config flow will fail and you'll need to repeat it
* In all other scenarios, the ADB key should be pre-approved and that timeout has no effect, so it doesn't matter if it's 10 or 30

It's not a big deal, though.

Reply:
May be I'm wrong, but I was thinking about a situation where the file with `adbkey` is physically deleted from HA folder and HA automatically generate a new one at restart. At this point the key is changed and so the device will pop-up again  for authorization, doesn't it?
Anyway timeout is already as described by you, 30 seconds for config flow, 10 seconds in all other scenario, that should not matter because key should be pre-approved.
But as you said, this is not a big deal, if you think that is better to have 30 sec, in all scenario I will change it for all, I leave final decision on your hand.😉

Reply:
I would use 30 seconds for all scenarios.

Reply:
Set timeout 30 seconds for all scenarios (default value)